// next/dist/next-server/server/next-server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const compression_1 = __importDefault(require("compression"));
const fs_1 = __importDefault(require("fs"));
const http_proxy_1 = __importDefault(require("http-proxy"));
const index_js_1 = __importDefault(require("next/dist/compiled/nanoid/index.js"));
const path_1 = require("path");
const querystring_1 = require("querystring");
const url_1 = require("url");
const check_custom_routes_1 = require("../../lib/check-custom-routes");
const coalesced_function_1 = require("../../lib/coalesced-function");
const constants_1 = require("../lib/constants");
const utils_1 = require("../lib/router/utils");
const envConfig = __importStar(require("../lib/runtime-config"));
const utils_2 = require("../lib/utils");
const api_utils_1 = require("./api-utils");
const config_1 = __importStar(require("./config"));
const path_match_1 = __importDefault(require("./lib/path-match"));
const recursive_readdir_sync_1 = require("./lib/recursive-readdir-sync");
const load_components_1 = require("./load-components");
const normalize_page_path_1 = require("./normalize-page-path");
const render_1 = require("./render");
const require_1 = require("./require");
const router_1 = __importStar(require("./router"));
const send_html_1 = require("./send-html");
const serve_static_1 = require("./serve-static");
const spr_cache_1 = require("./spr-cache");
const utils_3 = require("./utils");
const getCustomRouteMatcher = path_match_1.default(true);
class Server {
    constructor({ dir = '.', staticMarkup = false, quiet = false, conf = null, dev = false, } = {}) {
        this.dir = path_1.resolve(dir);
        this.quiet = quiet;
        const phase = this.currentPhase();
        this.nextConfig = config_1.default(phase, this.dir, conf);
        this.distDir = path_1.join(this.dir, this.nextConfig.distDir);
        this.publicDir = path_1.join(this.dir, constants_1.CLIENT_PUBLIC_FILES_PATH);
        this.hasStaticDir = fs_1.default.existsSync(path_1.join(this.dir, 'static'));
        // Only serverRuntimeConfig needs the default
        // publicRuntimeConfig gets it's default in client/index.js
        const { serverRuntimeConfig = {}, publicRuntimeConfig, assetPrefix, generateEtags, compress, } = this.nextConfig;
        this.buildId = this.readBuildId();
        this.renderOpts = {
            poweredByHeader: this.nextConfig.poweredByHeader,
            canonicalBase: this.nextConfig.amp.canonicalBase,
            documentMiddlewareEnabled: this.nextConfig.experimental
                .documentMiddleware,
            hasCssMode: this.nextConfig.experimental.css,
            staticMarkup,
            buildId: this.buildId,
            generateEtags,
            pages404: this.nextConfig.experimental.pages404,
        };
        // Only the `publicRuntimeConfig` key is exposed to the client side
        // It'll be rendered as part of __NEXT_DATA__ on the client side
        if (Object.keys(publicRuntimeConfig).length > 0) {
            this.renderOpts.runtimeConfig = publicRuntimeConfig;
        }
        if (compress && this.nextConfig.target === 'server') {
            this.compression = compression_1.default();
        }
        // Initialize next/config with the environment configuration
        envConfig.setConfig({
            serverRuntimeConfig,
            publicRuntimeConfig,
        });
        this.serverBuildDir = path_1.join(this.distDir, this._isLikeServerless ? constants_1.SERVERLESS_DIRECTORY : constants_1.SERVER_DIRECTORY);
        const pagesManifestPath = path_1.join(this.serverBuildDir, constants_1.PAGES_MANIFEST);
        if (!dev) {
            this.pagesManifest = require(pagesManifestPath);
        }
        this.router = new router_1.default(this.generateRoutes());
        this.setAssetPrefix(assetPrefix);
        // call init-server middleware, this is also handled
        // individually in serverless bundles when deployed
        if (!dev && this.nextConfig.experimental.plugins) {
            const initServer = require(path_1.join(this.serverBuildDir, 'init-server.js'))
                .default;
            this.onErrorMiddleware = require(path_1.join(this.serverBuildDir, 'on-error-server.js')).default;
            initServer();
        }
        spr_cache_1.initializeSprCache({
            dev,
            distDir: this.distDir,
            pagesDir: path_1.join(this.distDir, this._isLikeServerless
                ? constants_1.SERVERLESS_DIRECTORY
                : `${constants_1.SERVER_DIRECTORY}/static/${this.buildId}`, 'pages'),
            flushToDisk: this.nextConfig.experimental.sprFlushToDisk,
        });
    }
    currentPhase() {
        return constants_1.PHASE_PRODUCTION_SERVER;
    }
    logError(err) {
        if (this.onErrorMiddleware) {
            this.onErrorMiddleware({ err });
        }
        if (this.quiet)
            return;
        // tslint:disable-next-line
        console.error(err);
    }
    async handleRequest(req, res, parsedUrl) {
        // Parse url if parsedUrl not provided
        if (!parsedUrl || typeof parsedUrl !== 'object') {
            const url = req.url;
            parsedUrl = url_1.parse(url, true);
        }
        // Parse the querystring ourselves if the user doesn't handle querystring parsing
        if (typeof parsedUrl.query === 'string') {
            parsedUrl.query = querystring_1.parse(parsedUrl.query);
        }
        if (parsedUrl.pathname.startsWith(this.nextConfig.experimental.basePath)) {
            // If replace ends up replacing the full url it'll be `undefined`, meaning we have to default it to `/`
            parsedUrl.pathname =
                parsedUrl.pathname.replace(this.nextConfig.experimental.basePath, '') || '/';
            req.url = req.url.replace(this.nextConfig.experimental.basePath, '');
        }
        res.statusCode = 200;
        try {
            return await this.run(req, res, parsedUrl);
        }
        catch (err) {
            this.logError(err);
            res.statusCode = 500;
            res.end('Internal Server Error');
        }
    }
    getRequestHandler() {
        return this.handleRequest.bind(this);
    }
    setAssetPrefix(prefix) {
        this.renderOpts.assetPrefix = prefix ? prefix.replace(/\/$/, '') : '';
    }
    // Backwards compatibility
    async prepare() { }
    // Backwards compatibility
    async close() { }
    setImmutableAssetCacheControl(res) {
        res.setHeader('Cache-Control', 'public, max-age=31536000, immutable');
    }
    getCustomRoutes() {
        return require(path_1.join(this.distDir, constants_1.ROUTES_MANIFEST));
    }
    getPreviewProps() {
        if (this._cachedPreviewProps) {
            return this._cachedPreviewProps;
        }
        return (this._cachedPreviewProps = require(path_1.join(this.distDir, constants_1.PRERENDER_MANIFEST)).preview);
    }
    generateRoutes() {
        this.customRoutes = this.getCustomRoutes();
        const publicRoutes = fs_1.default.existsSync(this.publicDir)
            ? this.generatePublicRoutes()
            : [];
        const staticFilesRoute = this.hasStaticDir
            ? [
                {
                    // It's very important to keep this route's param optional.
                    // (but it should support as many params as needed, separated by '/')
                    // Otherwise this will lead to a pretty simple DOS attack.
                    // See more: https://github.com/zeit/next.js/issues/2617
                    match: router_1.route('/static/:path*'),
                    name: 'static catchall',
                    fn: async (req, res, params, parsedUrl) => {
                        const p = path_1.join(this.dir, 'static', ...(params.path || []));
                        await this.serveStatic(req, res, p, parsedUrl);
                        return {
                            finished: true,
                        };
                    },
                },
            ]
            : [];
        let headers = [];
        let rewrites = [];
        let redirects = [];
        const fsRoutes = [
            {
                match: router_1.route('/_next/static/:path*'),
                type: 'route',
                name: '_next/static catchall',
                fn: async (req, res, params, parsedUrl) => {
                    // The commons folder holds commonschunk files
                    // The chunks folder holds dynamic entries
                    // The buildId folder holds pages and potentially other assets. As buildId changes per build it can be long-term cached.
                    // make sure to 404 for /_next/static itself
                    if (!params.path) {
                        await this.render404(req, res, parsedUrl);
                        return {
                            finished: true,
                        };
                    }
                    if (params.path[0] === constants_1.CLIENT_STATIC_FILES_RUNTIME ||
                        params.path[0] === 'chunks' ||
                        params.path[0] === 'css' ||
                        params.path[0] === 'media' ||
                        params.path[0] === this.buildId) {
                        this.setImmutableAssetCacheControl(res);
                    }
                    const p = path_1.join(this.distDir, constants_1.CLIENT_STATIC_FILES_PATH, ...(params.path || []));
                    await this.serveStatic(req, res, p, parsedUrl);
                    return {
                        finished: true,
                    };
                },
            },
            {
                match: router_1.route('/_next/data/:path*'),
                type: 'route',
                name: '_next/data catchall',
                fn: async (req, res, params, _parsedUrl) => {
                    // Make sure to 404 for /_next/data/ itself and
                    // we also want to 404 if the buildId isn't correct
                    if (!params.path || params.path[0] !== this.buildId) {
                        await this.render404(req, res, _parsedUrl);
                        return {
                            finished: true,
                        };
                    }
                    // remove buildId from URL
                    params.path.shift();
                    // show 404 if it doesn't end with .json
                    if (!params.path[params.path.length - 1].endsWith('.json')) {
                        await this.render404(req, res, _parsedUrl);
                        return {
                            finished: true,
                        };
                    }
                    // re-create page's pathname
                    const pathname = `/${params.path.join('/')}`
                        .replace(/\.json$/, '')
                        .replace(/\/index$/, '/');
                    req.url = pathname;
                    const parsedUrl = url_1.parse(pathname, true);
                    await this.render(req, res, pathname, Object.assign(Object.assign({}, _parsedUrl.query), { _nextDataReq: '1' }), parsedUrl);
                    return {
                        finished: true,
                    };
                },
            },
            {
                match: router_1.route('/_next/:path*'),
                type: 'route',
                name: '_next catchall',
                // This path is needed because `render()` does a check for `/_next` and the calls the routing again
                fn: async (req, res, _params, parsedUrl) => {
                    await this.render404(req, res, parsedUrl);
                    return {
                        finished: true,
                    };
                },
            },
            ...publicRoutes,
            ...staticFilesRoute,
        ];
        if (this.customRoutes) {
            const getCustomRoute = (r, type) => (Object.assign(Object.assign({}, r), { type, match: getCustomRouteMatcher(r.source), name: type, fn: async (req, res, params, parsedUrl) => ({ finished: false }) }));
            // Headers come very first
            headers = this.customRoutes.headers.map(r => {
                const route = getCustomRoute(r, 'header');
                return {
                    match: route.match,
                    type: route.type,
                    name: `${route.type} ${route.source} header route`,
                    fn: async (_req, res, _params, _parsedUrl) => {
                        for (const header of route.headers) {
                            res.setHeader(header.key, header.value);
                        }
                        return { finished: false };
                    },
                };
            });
            redirects = this.customRoutes.redirects.map(redirect => {
                const route = getCustomRoute(redirect, 'redirect');
                return {
                    type: route.type,
                    match: route.match,
                    statusCode: route.statusCode,
                    name: `Redirect route`,
                    fn: async (_req, res, params, _parsedUrl) => {
                        const { parsedDestination } = router_1.prepareDestination(route.destination, params);
                        const updatedDestination = url_1.format(parsedDestination);
                        res.setHeader('Location', updatedDestination);
                        res.statusCode = check_custom_routes_1.getRedirectStatus(route);
                        // Since IE11 doesn't support the 308 header add backwards
                        // compatibility using refresh header
                        if (res.statusCode === 308) {
                            res.setHeader('Refresh', `0;url=${updatedDestination}`);
                        }
                        res.end();
                        return {
                            finished: true,
                        };
                    },
                };
            });
            rewrites = this.customRoutes.rewrites.map(rewrite => {
                const route = getCustomRoute(rewrite, 'rewrite');
                return {
                    check: true,
                    type: route.type,
                    name: `Rewrite route`,
                    match: route.match,
                    fn: async (req, res, params, _parsedUrl) => {
                        const { newUrl, parsedDestination } = router_1.prepareDestination(route.destination, params);
                        // external rewrite, proxy it
                        if (parsedDestination.protocol) {
                            const target = url_1.format(parsedDestination);
                            const proxy = new http_proxy_1.default({
                                target,
                                changeOrigin: true,
                                ignorePath: true,
                            });
                            proxy.web(req, res);
                            proxy.on('error', (err) => {
                                console.error(`Error occurred proxying ${target}`, err);
                            });
                            return {
                                finished: true,
                            };
                        }
                        ;
                        req._nextDidRewrite = true;
                        return {
                            finished: false,
                            pathname: newUrl,
                            query: parsedDestination.query,
                        };
                    },
                };
            });
        }
        const catchAllRoute = {
            match: router_1.route('/:path*'),
            type: 'route',
            name: 'Catchall render',
            fn: async (req, res, params, parsedUrl) => {
                var _a, _b;
                const { pathname, query } = parsedUrl;
                if (!pathname) {
                    throw new Error('pathname is undefined');
                }
                if (((_b = (_a = params) === null || _a === void 0 ? void 0 : _a.path) === null || _b === void 0 ? void 0 : _b[0]) === 'api') {
                    const handled = await this.handleApiRequest(req, res, pathname, query);
                    if (handled) {
                        return { finished: true };
                    }
                }
                await this.render(req, res, pathname, query, parsedUrl);
                return {
                    finished: true,
                };
            },
        };
        const { useFileSystemPublicRoutes } = this.nextConfig;
        if (useFileSystemPublicRoutes) {
            this.dynamicRoutes = this.getDynamicRoutes();
        }
        return {
            headers,
            fsRoutes,
            rewrites,
            redirects,
            catchAllRoute,
            useFileSystemPublicRoutes,
            dynamicRoutes: this.dynamicRoutes,
            pageChecker: this.hasPage.bind(this),
        };
    }
    async getPagePath(pathname) {
        return require_1.getPagePath(pathname, this.distDir, this._isLikeServerless, this.renderOpts.dev);
    }
    async hasPage(pathname) {
        let found = false;
        try {
            found = !!(await this.getPagePath(pathname));
        }
        catch (_) { }
        return found;
    }
    async _beforeCatchAllRender(_req, _res, _params, _parsedUrl) {
        return false;
    }
    // Used to build API page in development
    async ensureApiPage(pathname) { }
    /**
     * Resolves `API` request, in development builds on demand
     * @param req http request
     * @param res http response
     * @param pathname path of request
     */
    async handleApiRequest(req, res, pathname, query) {
        let page = pathname;
        let params = false;
        let pageFound = await this.hasPage(page);
        if (!pageFound && this.dynamicRoutes) {
            for (const dynamicRoute of this.dynamicRoutes) {
                params = dynamicRoute.match(pathname);
                if (dynamicRoute.page.startsWith('/api') && params) {
                    page = dynamicRoute.page;
                    pageFound = true;
                    break;
                }
            }
        }
        if (!pageFound) {
            return false;
        }
        // Make sure the page is built before getting the path
        // or else it won't be in the manifest yet
        await this.ensureApiPage(page);
        const builtPagePath = await this.getPagePath(page);
        const pageModule = require(builtPagePath);
        query = Object.assign(Object.assign({}, query), params);
        if (!this.renderOpts.dev && this._isLikeServerless) {
            if (typeof pageModule.default === 'function') {
                this.prepareServerlessUrl(req, query);
                await pageModule.default(req, res);
                return true;
            }
        }
        const previewProps = this.getPreviewProps();
        await api_utils_1.apiResolver(req, res, query, pageModule, Object.assign({}, previewProps), this.onErrorMiddleware);
        return true;
    }
    generatePublicRoutes() {
        const publicFiles = new Set(recursive_readdir_sync_1.recursiveReadDirSync(this.publicDir).map(p => p.replace(/\\/g, '/')));
        return [
            {
                match: router_1.route('/:path*'),
                name: 'public folder catchall',
                fn: async (req, res, params, parsedUrl) => {
                    const path = `/${(params.path || []).join('/')}`;
                    if (publicFiles.has(path)) {
                        await this.serveStatic(req, res, 
                        // we need to re-encode it since send decodes it
                        path_1.join(this.dir, 'public', encodeURIComponent(path)), parsedUrl);
                        return {
                            finished: true,
                        };
                    }
                    return {
                        finished: false,
                    };
                },
            },
        ];
    }
    getDynamicRoutes() {
        const dynamicRoutedPages = Object.keys(this.pagesManifest).filter(utils_1.isDynamicRoute);
        return utils_1.getSortedRoutes(dynamicRoutedPages).map(page => ({
            page,
            match: utils_1.getRouteMatcher(utils_1.getRouteRegex(page)),
        }));
    }
    handleCompression(req, res) {
        if (this.compression) {
            this.compression(req, res, () => { });
        }
    }
    async run(req, res, parsedUrl) {
        this.handleCompression(req, res);
        try {
            const matched = await this.router.execute(req, res, parsedUrl);
            if (matched) {
                return;
            }
        }
        catch (err) {
            if (err.code === 'DECODE_FAILED') {
                res.statusCode = 400;
                return this.renderError(null, req, res, '/_error', {});
            }
            throw err;
        }
        await this.render404(req, res, parsedUrl);
    }
    async sendHTML(req, res, html) {
        const { generateEtags, poweredByHeader } = this.renderOpts;
        return send_html_1.sendHTML(req, res, html, { generateEtags, poweredByHeader });
    }
    async render(req, res, pathname, query = {}, parsedUrl) {
        const url = req.url;
        if (url.match(/^\/_next\//) ||
            (this.hasStaticDir && url.match(/^\/static\//))) {
            return this.handleRequest(req, res, parsedUrl);
        }
        if (utils_3.isBlockedPage(pathname)) {
            return this.render404(req, res, parsedUrl);
        }
        const html = await this.renderToHTML(req, res, pathname, query, {});
        // Request was ended by the user
        if (html === null) {
            return;
        }
        return this.sendHTML(req, res, html);
    }
    async findPageComponents(pathname, query = {}, params = null) {
        const paths = [
            // try serving a static AMP version first
            query.amp ? normalize_page_path_1.normalizePagePath(pathname) + '.amp' : null,
            pathname,
        ].filter(Boolean);
        for (const pagePath of paths) {
            try {
                const components = await load_components_1.loadComponents(this.distDir, this.buildId, pagePath, !this.renderOpts.dev && this._isLikeServerless);
                return {
                    components,
                    query: Object.assign(Object.assign({}, (components.unstable_getStaticProps
                        ? { _nextDataReq: query._nextDataReq }
                        : query)), (params || {})),
                };
            }
            catch (err) {
                if (err.code !== 'ENOENT')
                    throw err;
            }
        }
        return null;
    }
    __sendPayload(res, payload, type, revalidate) {
        // TODO: ETag? Cache-Control headers? Next-specific headers?
        res.setHeader('Content-Type', type);
        res.setHeader('Content-Length', Buffer.byteLength(payload));
        if (!this.renderOpts.dev) {
            if (revalidate) {
                res.setHeader('Cache-Control', revalidate < 0
                    ? `no-cache, no-store, must-revalidate`
                    : `s-maxage=${revalidate}, stale-while-revalidate`);
            }
            else if (revalidate === false) {
                res.setHeader('Cache-Control', `s-maxage=31536000, stale-while-revalidate`);
            }
        }
        res.end(payload);
    }
    prepareServerlessUrl(req, query) {
        const curUrl = url_1.parse(req.url, true);
        req.url = url_1.format(Object.assign(Object.assign({}, curUrl), { search: undefined, query: Object.assign(Object.assign({}, curUrl.query), query) }));
    }
    async renderToHTMLWithComponents(req, res, pathname, { components, query }, opts) {
        var _a, _b;
        // we need to ensure the status code if /404 is visited directly
        if (this.nextConfig.experimental.pages404 && pathname === '/404') {
            res.statusCode = 404;
        }
        // handle static page
        if (typeof components.Component === 'string') {
            return components.Component;
        }
        // check request state
        const isLikeServerless = typeof components.Component === 'object' &&
            typeof components.Component.renderReqToHTML === 'function';
        const isSSG = !!components.unstable_getStaticProps;
        const isServerProps = !!components.unstable_getServerProps;
        // Toggle whether or not this is a Data request
        const isDataReq = query._nextDataReq;
        delete query._nextDataReq;
        // Serverless requests need its URL transformed back into the original
        // request path (to emulate lambda behavior in production)
        if (isLikeServerless && isDataReq) {
            let { pathname } = url_1.parse(req.url || '', true);
            pathname = !pathname || pathname === '/' ? '/index' : pathname;
            req.url = url_1.format({
                pathname: `/_next/data/${this.buildId}${pathname}.json`,
                query,
            });
        }
        // non-spr requests should render like normal
        if (!isSSG) {
            // handle serverless
            if (isLikeServerless) {
                if (isDataReq) {
                    const renderResult = await components.Component.renderReqToHTML(req, res, true);
                    this.__sendPayload(res, JSON.stringify((_b = (_a = renderResult) === null || _a === void 0 ? void 0 : _a.renderOpts) === null || _b === void 0 ? void 0 : _b.pageData), 'application/json', -1);
                    return null;
                }
                this.prepareServerlessUrl(req, query);
                return components.Component.renderReqToHTML(req, res);
            }
            if (isDataReq && isServerProps) {
                const props = await render_1.renderToHTML(req, res, pathname, query, Object.assign(Object.assign(Object.assign({}, components), opts), { isDataReq }));
                this.__sendPayload(res, JSON.stringify(props), 'application/json', -1);
                return null;
            }
            return render_1.renderToHTML(req, res, pathname, query, Object.assign(Object.assign({}, components), opts));
        }
        const previewProps = this.getPreviewProps();
        const previewData = api_utils_1.tryGetPreviewData(req, res, Object.assign({}, previewProps));
        const isPreviewMode = previewData !== false;
        // Compute the SPR cache key
        const ssgCacheKey = isPreviewMode
            ? `__` + index_js_1.default() // Preview mode uses a throw away key to not coalesce preview invokes
            : url_1.parse(req.url || '').pathname;
        // Complete the response with cached data if its present
        const cachedData = isPreviewMode
            ? // Preview data bypasses the cache
                undefined
            : await spr_cache_1.getSprCache(ssgCacheKey);
        if (cachedData) {
            const data = isDataReq
                ? JSON.stringify(cachedData.pageData)
                : cachedData.html;
            this.__sendPayload(res, data, isDataReq ? 'application/json' : 'text/html; charset=utf-8', cachedData.curRevalidate);
            // Stop the request chain here if the data we sent was up-to-date
            if (!cachedData.isStale) {
                return null;
            }
        }
        // If we're here, that means data is missing or it's stale.
        const doRender = coalesced_function_1.withCoalescedInvoke(async function () {
            let pageData;
            let html;
            let sprRevalidate;
            let renderResult;
            // handle serverless
            if (isLikeServerless) {
                renderResult = await components.Component.renderReqToHTML(req, res, true);
                html = renderResult.html;
                pageData = renderResult.renderOpts.pageData;
                sprRevalidate = renderResult.renderOpts.revalidate;
            }
            else {
                const renderOpts = Object.assign(Object.assign({}, components), opts);
                renderResult = await render_1.renderToHTML(req, res, pathname, query, renderOpts);
                html = renderResult;
                pageData = renderOpts.pageData;
                sprRevalidate = renderOpts.revalidate;
            }
            return { html, pageData, sprRevalidate };
        });
        const isProduction = !this.renderOpts.dev;
        const isDynamicPathname = utils_1.isDynamicRoute(pathname);
        const didRespond = utils_2.isResSent(res);
        // const isForcedBlocking =
        //   req.headers['X-Prerender-Bypass-Mode'] !== 'Blocking'
        // When we did not respond from cache, we need to choose to block on
        // rendering or return a skeleton.
        //
        // * Data requests always block.
        //
        // * Preview mode toggles all pages to be resolved in a blocking manner.
        //
        // * Non-dynamic pages should block (though this is an be an impossible
        //   case in production).
        //
        // * Dynamic pages should return their skeleton, then finish the data
        //   request on the client-side.
        //
        if (!didRespond &&
            !isDataReq &&
            !isPreviewMode &&
            isDynamicPathname &&
            // TODO: development should trigger fallback when the path is not in
            // `getStaticPaths`, for now, let's assume it is.
            isProduction) {
            let html;
            // Production already emitted the fallback as static HTML.
            if (isProduction) {
                html = await spr_cache_1.getFallback(pathname);
            }
            // We need to generate the fallback on-demand for development.
            else {
                query.__nextFallback = 'true';
                if (isLikeServerless) {
                    this.prepareServerlessUrl(req, query);
                    html = await components.Component.renderReqToHTML(req, res);
                }
                else {
                    html = (await render_1.renderToHTML(req, res, pathname, query, Object.assign(Object.assign({}, components), opts)));
                }
            }
            this.__sendPayload(res, html, 'text/html; charset=utf-8');
        }
        const { isOrigin, value: { html, pageData, sprRevalidate }, } = await doRender(ssgCacheKey, []);
        if (!utils_2.isResSent(res)) {
            this.__sendPayload(res, isDataReq ? JSON.stringify(pageData) : html, isDataReq ? 'application/json' : 'text/html; charset=utf-8', sprRevalidate);
        }
        // Update the SPR cache if the head request
        if (isOrigin) {
            // Preview mode should not be stored in cache
            if (!isPreviewMode) {
                await spr_cache_1.setSprCache(ssgCacheKey, { html: html, pageData }, sprRevalidate);
            }
        }
        return null;
    }
    async renderToHTML(req, res, pathname, query = {}, { amphtml, hasAmp, } = {}) {
        try {
            const result = await this.findPageComponents(pathname, query);
            if (result) {
                return await this.renderToHTMLWithComponents(req, res, pathname, result, Object.assign(Object.assign({}, this.renderOpts), { amphtml, hasAmp }));
            }
            if (this.dynamicRoutes) {
                for (const dynamicRoute of this.dynamicRoutes) {
                    const params = dynamicRoute.match(pathname);
                    if (!params) {
                        continue;
                    }
                    const result = await this.findPageComponents(dynamicRoute.page, query, params);
                    if (result) {
                        return await this.renderToHTMLWithComponents(req, res, dynamicRoute.page, result, Object.assign(Object.assign({}, this.renderOpts), { params,
                            amphtml,
                            hasAmp }));
                    }
                }
            }
        }
        catch (err) {
            this.logError(err);
            res.statusCode = 500;
            return await this.renderErrorToHTML(err, req, res, pathname, query);
        }
        res.statusCode = 404;
        return await this.renderErrorToHTML(null, req, res, pathname, query);
    }
    async renderError(err, req, res, pathname, query = {}) {
        res.setHeader('Cache-Control', 'no-cache, no-store, max-age=0, must-revalidate');
        const html = await this.renderErrorToHTML(err, req, res, pathname, query);
        if (html === null) {
            return;
        }
        return this.sendHTML(req, res, html);
    }
    async renderErrorToHTML(err, req, res, _pathname, query = {}) {
        let result = null;
        const { static404, pages404 } = this.nextConfig.experimental;
        const is404 = res.statusCode === 404;
        let using404Page = false;
        // use static 404 page if available and is 404 response
        if (is404) {
            if (static404) {
                result = await this.findPageComponents('/_errors/404');
            }
            // use 404 if /_errors/404 isn't available which occurs
            // during development and when _app has getInitialProps
            if (!result && pages404) {
                result = await this.findPageComponents('/404');
                using404Page = result !== null;
            }
        }
        if (!result) {
            result = await this.findPageComponents('/_error', query);
        }
        let html;
        try {
            html = await this.renderToHTMLWithComponents(req, res, using404Page ? '/404' : '/_error', result, Object.assign(Object.assign({}, this.renderOpts), { err }));
        }
        catch (err) {
            console.error(err);
            res.statusCode = 500;
            html = 'Internal Server Error';
        }
        return html;
    }
    async render404(req, res, parsedUrl) {
        const url = req.url;
        const { pathname, query } = parsedUrl ? parsedUrl : url_1.parse(url, true);
        if (!pathname) {
            throw new Error('pathname is undefined');
        }
        res.statusCode = 404;
        return this.renderError(null, req, res, pathname, query);
    }
    async serveStatic(req, res, path, parsedUrl) {
        if (!this.isServeableUrl(path)) {
            return this.render404(req, res, parsedUrl);
        }
        if (!(req.method === 'GET' || req.method === 'HEAD')) {
            res.statusCode = 405;
            res.setHeader('Allow', ['GET', 'HEAD']);
            return this.renderError(null, req, res, path);
        }
        try {
            await serve_static_1.serveStatic(req, res, path);
        }
        catch (err) {
            if (err.code === 'ENOENT' || err.statusCode === 404) {
                this.render404(req, res, parsedUrl);
            }
            else if (err.statusCode === 412) {
                res.statusCode = 412;
                return this.renderError(err, req, res, path);
            }
            else {
                throw err;
            }
        }
    }
    isServeableUrl(path) {
        const resolved = path_1.resolve(path);
        if (resolved.indexOf(path_1.join(this.distDir) + path_1.sep) !== 0 &&
            resolved.indexOf(path_1.join(this.dir, 'static') + path_1.sep) !== 0 &&
            resolved.indexOf(path_1.join(this.dir, 'public') + path_1.sep) !== 0) {
            // Seems like the user is trying to traverse the filesystem.
            return false;
        }
        return true;
    }
    readBuildId() {
        const buildIdFile = path_1.join(this.distDir, constants_1.BUILD_ID_FILE);
        try {
            return fs_1.default.readFileSync(buildIdFile, 'utf8').trim();
        }
        catch (err) {
            if (!fs_1.default.existsSync(buildIdFile)) {
                throw new Error(`Could not find a valid build in the '${this.distDir}' directory! Try building your app with 'next build' before starting the server.`);
            }
            throw err;
        }
    }
    get _isLikeServerless() {
        return config_1.isTargetLikeServerless(this.nextConfig.target);
    }
}
exports.default = Server;
